File: basic_python/practices/01. Array & Hashing/Group Anagrams/group_anagrams.cpp
#include <iostream>
#include <vector>
#include <string>
#include <map>
#include <algorithm>
#include <unordered_map>
using namespace std;

class GroupAnagrams {
public:
    vector<vector<string>> usingSort(vector<string>& strs) {
        map<string, vector<string>> res;
        for (const string& s : strs) {
            string sorted_s = s;
            sort(sorted_s.begin(), sorted_s.end());
            res[sorted_s].push_back(s);
        }
        
        vector<vector<string>> result;
        for (auto& pair : res) {
            result.push_back(pair.second);
        }
        return result;
    }
    
    vector<vector<string>> usingHash(vector<string>& strs) {
        unordered_map<string, vector<string>> res;
        for (const string& s : strs) {
            vector<int> count(26, 0);
            for (char c : s) {
                count[c - 'a']++;
            }
            
            // Convert count to string key
            string key;
            for (int num : count) {
                key += to_string(num) + ",";
            }
            res[key].push_back(s);
        }
        
        vector<vector<string>> result;
        for (auto& pair : res) {
            result.push_back(pair.second);
        }
        return result;
    }
};

int main() {
    cout << "Enter your list of strings (type one word per line):" << endl;
    int n_words;
    cout << "Number of words: ";
    cin >> n_words;
    cin.ignore(); // Clear newline character
    
    vector<string> raw;
    for (int i = 0; i < n_words; i++) {
        string vocab;
        cout << "\t " << (1 + i) << ": \t";
        getline(cin, vocab);
        raw.push_back(vocab);
    }
    
    string method;
    cout << "Choose method: 'sort' or 'hash': ";
    getline(cin, method);
    transform(method.begin(), method.end(), method.begin(), ::tolower);
    
    GroupAnagrams solver;
    vector<vector<string>> result;
    
    if (method.find("hash") != string::npos) {
        result = solver.usingHash(raw);
    } else {
        result = solver.usingSort(raw);
    }
    
    cout << "Grouped anagrams: ";
    for (const auto& group : result) {
        cout << "[";
        for (size_t i = 0; i < group.size(); i++) {
            cout << "\"" << group[i] << "\"";
            if (i < group.size() - 1) cout << ", ";
        }
        cout << "] ";
    }
    cout << endl;
    
    return 0;
}

File: basic_python/practices/Data Structures/01. Array & Hashing/Group Anagrams/group_anagrams.py
from collections import defaultdict

class groupAnagrams:
    def using_sort(self, strs: list[str]) -> list[list[str]]:
        res = defaultdict(list)
        for s in strs:
            sorted_str = ''.join(sorted(s))
            res[sorted_str].append(s)
        return list(res.values())
    
    def using_hash(self, strs: list[str]) -> list[list[str]]:
        res = defaultdict(list)
        for s in strs:
            count = [0] * 26
            for c in s:
                count[ord(c) - ord('a')] += 1
            res[tuple(count)].append(s)
        return list(res.values())

if __name__ == "__main__":
    print("Enter your list of strings (type one word per line):")
    n_words = int(input("Number of words: "))
    
    raw = []
    for _ in range(n_words):
        vocab = input(f"\t {1 + _}: \t")
        raw.append(vocab)

    method = input("Choose method: 'sort' or 'hash': ").strip().lower()

    solver = groupAnagrams()

    if method == "hash":
        result = solver.using_hash(raw)
    else:
        result = solver.using_sort(raw)

    print("Grouped anagrams:", result)